nfs/src/rpc_task.cpp

```diff
@@ -1313,30 +1273,49 @@ void rpc_task::run_write()
      * membufs. We don't wait for the writes to actually finish, which means
      * we support buffered writes.
      */
-    std::vector<bytes_chunk> bc_vec =
-            copy_to_cache(this, ino, bufv, offset, length, error_code);
+    uint64_t periodic_bytes = 0;
+
+    copy_to_cache(this, ino, bufv, offset, length, error_code, &extent_left, &extent_right);
     if (error_code != 0) {
         AZLogWarn("[{}] copy_to_cache failed with error={}, "
                   "failing write!", ino, error_code);
         reply_error(error_code);
         return;
     }
 
-    assert(bc_vec.size() > 0);
+    assert(extent_right >= extent_left);
 
-    for (bytes_chunk& bc : bc_vec) {
-        /*
-         * Create the flush task to carry out the write.
-         */
-        struct rpc_task *flush_task =
-            get_client()->get_rpc_task_helper()->alloc_rpc_task(FUSE_FLUSH);
-        flush_task->init_flush(nullptr /* fuse_req */, ino);
+    /*
+     * Get the prune goals, if it's more than data already in-flight
+     * data to BLOB, then issue fresh flush.
+     */
+    inode->filecache_handle->get_prune_goals(nullptr, &periodic_bytes);
+    AZLogDebug("periodic_bytes: {}, extent_left{}, extent_right: {}",periodic_bytes, extent_left, extent_right);
+
+    if (inode->inflight_bytes >= periodic_bytes) {
+        AZLogDebug("Reply write without syncing to Blob");
+        reply_write(length);
+        return;
+    }
```

Comment:
Later when we are doing exact block sized writes, then waiting for dirty data to accumulate more than the block size is not useful, instead it'll be detrimental as we would be delaying writes. This comment is just to make sure we are on the same page.

One can argue whether we need to wait even now, i.e., why don't we send out a write rpc as soon as "extent_right-extent_left" tells us that the current dirty extent is >= wsize. I think that'll perform equally fine with the current server write scheduler, since the application writes are happening very fast, so even if we send right away we will get good merging on the server. This is something we can experiment and see.

Reply:
Infact I'm thinking that copy_to_cache should tell the caller if it must flush now. So all the logic of flushing detection can be inside copy_to_cache and the caller just needs to honor that.